packages/@lwc/ssr-runtime/src/styles.ts

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { Stylesheets, Stylesheet } from '@lwc/engine-core';
+import { validateStyleTextContents } from './validate-style-text-contents';
+import type { LightningElementConstructor } from './lightning-element';
+
+function flattenStylesheets(stylesheets: Stylesheets): Array<Stylesheet> {
+    // recursively flatten
+    // @ts-expect-error we can optimize this function later
+    return stylesheets.filter(Boolean).flat(Infinity);
+}
+
+export function hasScopedStaticStylesheets(Component: LightningElementConstructor): boolean {
+    const { stylesheets } = Component;
+    if (stylesheets) {
+        return flattenStylesheets(stylesheets).some((stylesheet: any) => stylesheet.$scoped$);
+    }
+    return false;
+}
+
+export function renderStylesheets(
+    stylesheets: Stylesheets,
+    scopeToken: string,
+    Component: LightningElementConstructor,
+    hasScopedTemplateStyles: boolean
+): string {
+    const hasAnyScopedStyles = hasScopedTemplateStyles || hasScopedStaticStylesheets(Component);
+
+    let result = '';
+
+    for (const stylesheet of flattenStylesheets(stylesheets)) {
```

Comment:
Do we want to be fancy? Could change `flattenStylesheets` to a generator function!

Reply:
I think we're going to be getting rid of `yield` eventually anyway